share/shwap/p2p/bitswap/block_fetch.go

```diff
@@ -0,0 +1,143 @@
+		// store the PopulateFn s.t. hasher can access it
+		// and fill in the Block
+		_, exists := populators.LoadOrStore(idStr, p)
+		if exists {
+			// in case there is ongoing fetch happening for the same Block elsewhere
+			// and PopulateFn has already been set -- mark the Block as duplicate
+			duplicate[cid] = blk
+		} else {
+			// only do the cleanup if we stored the PopulateFn
+			defer populators.Delete(idStr)
+		}
+	}
+
+	blkCh, err := fetcher.GetBlocks(ctx, cids)
+	if err != nil {
+		return fmt.Errorf("fetching bitswap blocks: %w", err)
+	}
+
+	for blk := range blkCh { // GetBlocks closes blkCh on ctx cancellation
+		// check if the blk is a duplicate
+		id, ok := duplicate[blk.Cid()]
+		if !ok {
+			continue
+		}
+		// if it is, we have to populate it ourselves instead of hasher,
+		// as there is only one PopulateFN allowed per ID
+		err := id.Populate(root)(blk.RawData())
+		if err != nil {
+			// this case should never happen in practice
+			// and if so something is really wrong
+			panic(fmt.Sprintf("populating duplicate block: %s", err))
+		}
+		// NOTE: This approach has a downside that we redo deserialization and computationally
+		// expensive computation for as many duplicates. We tried solutions that doesn't have this
+		// problem, but they are *much* more complex. Considering this a rare edge-case the tradeoff
+		// towards simplicity has been made.
+	}
+
+	return ctx.Err()
+}
+
+// populators exists to communicate between Fetch and hasher.
+//
+// Fetch registers PopulateFNs that hasher then uses to validate and populate Block responses coming
+// through Bitswap
+//
+// Bitswap does not provide *stateful* verification out of the box and by default
+// messages are verified by their respective MultiHashes that are registered globally.
+// For every Block type there is a global hasher registered that accesses stored PopulateFn once a
+// message is received. It then uses PopulateFn to validate and fill in the respective Block
+//
+// sync.Map is used to minimize contention for disjoint keys
+var populators sync.Map
+
+// hasher implements hash.Hash to be registered as custom multihash
+// hasher is the *hack* to inject custom verification logic into Bitswap
+type hasher struct {
+	// IDSize of the respective Shwap container
+	IDSize int // to be set during hasher registration
+
+	sum []byte
+}
+
+func (h *hasher) Write(data []byte) (int, error) {
+	const pbOffset = 2 // this assumes the protobuf serialization is in use
+	if len(data) < h.IDSize+pbOffset {
+		err := fmt.Errorf("shwap/bitswap hasher: insufficient data size")
+		log.Error()
+		return 0, err
+	}
+	// extract Block out of data
+	// we do this on the raw data to:
+	//  * Avoid complicating hasher with generalized bytes -> type unmarshalling
+	//  * Avoid type allocations
+	id := data[pbOffset : h.IDSize+pbOffset]
+	// get registered PopulateFn and use it to check data validity and
+	// pass it to Fetch caller
+	val, ok := populators.Load(string(id))
```

Comment:
> Otherwuse same populator might be triggered more than 1 time on same container.

This case is not possible in practice because for it to happen, the time between [the bitswap chan serving](https://github.com/celestiaorg/celestia-node/pull/3421/files#diff-07d853bb5fe2d749a2d35eaf7dee1d6920407bee44d6c8eff387f80a80a7b97dR50) and [the cleanup](https://github.com/celestiaorg/celestia-node/pull/3421/files#diff-07d853bb5fe2d749a2d35eaf7dee1d6920407bee44d6c8eff387f80a80a7b97dR41) should be bigger than the time it takes for Bitswap to fetch the block(again). However, I like the idea of using `LoadAndDelete`, and I couldn't use it in the iteration where there were multiple populatorFN registered per ID, with single populatorFn this may work

Reply:
So I tried, and the duplicate test became very flaky. I will debug later